### ISSUE TITLE
textform vs wordform

### ISSUE THREAD

**Issue by sylvainkahane** [UniversalDependencies/docs#683]
We had a discussion with @kimgerdes  and @bguil  today about many problems concerning requests on wordforms.

Currently there is only one form for a word which is the form in the text (I call it the textform). But sometimes this form is not convenient and it is not the form a linguist wants to request. Here are some cases:
* the word is at the beginning of the sentence and is written with a capital (_Did you know that?_)
* the whole word is written with capitals (_YEAHHH_)
* many characters exist for the “same” symbol,  such as apostrophes or quotes (" vs “, cf. Fr. _aujourd'hui_ vs _aujourd’hui_, or … (one character) vs ... (three characters))
* there is a spelling errors (_wich_ for _which_) including subtypes:
1. diacritics or special characters are forgotten (Fr. _ca_ for _ça_, Fr. _coeur_ for _cœur_)
2. the word is split (_egg plant_ for _eggplant_)
* diacritics are specially problematic with capitals (Fr. _A_ for _à_)
* there is an amalgam (_doesn’t_ = _does not_, Fr. _du_ = _de le_)

For all these problems we propose a simple and single solution: to introduce two forms, textform and wordform.
Here are the solutions for our examples:

```
id	textform	wordform	lemma
1	Did		did		do

id	textform	wordform	lemma
1	YEAHHH	yeahhh	yeah

id	textform	wordform
1	A		à

id	textform	wordform	lemma
1	doesn't	        does	        do
2	_		not		not

id	textform	wordform
1	du		de
2	_		le

id	textform	wordform	relation
1	egg		eggplant
2	plant	_		       goeswith
```
(A validation rule: wordform=_ iff relation=goeswith.)

The textform is the form in the text that we cannot change. We need it to make the mapping between the syntactic annotation and the text. The wordform is the (inflected) form, which is the form associated to the word by a linguist. This form can have been modified for several reason: strange orthographical convention that requests an uppercase at the beginning of a sentence, spelling errors, amalgams and portemanteau form.

Note that with this we solve many problems of tokenisation (see #641 ): we don't need to discuss whether En. _wanna_ is split in _wan na_ or _want to_. A (text)word is never split.
```
id	textform	wordform
1	wanna		want
2	_		to
```

**Comment by nschneid** [comment]
Thanks for the suggestion. I've been wondering lately whether there should be an easy way to extract a parsed version of each sentence that incorporates orthographic normalization on the wordforms and merges `goeswith`-connected tokens. This proposal would enable that (an alternative would be a conversion script).

Why not simply eliminate the `goeswith` relation in the "eggplant" example? `goeswith` is not a real syntactic relation, just an orthographic relation that seems redundant if textform is separated from wordform.

Note that orthographic features described [here](https://universaldependencies.org/u/overview/typos.html) already allow spelling errors in the token to be corrected, but I am not sure a newcomer who wants to use the data in a normalized form would find it obvious which transformations to apply.

**Comment by sylvainkahane** [comment]
About `goeswith`: we need it at the text level. The text is what a parser see. The parser tokenizes the text into textforms and links them to obtain a dependency tree. So `goeswith` is useful at this level. The wordforms appear at the second level, the "linguistic syntactic level".

An additional comment. We don't want empty nodes in parsing and the parser won't consider empty textforms. It is only at the linguistic syntactic level that some textforms can be split and empty wordforms can be introduced. It is generally not possible to tokenize into wordforms before the syntactic parsing. For instance, in French _des_ can be an indefinite DET (_je lis des livres_ 'I read (DET) books') or an amalgam ADP+DET _de+les_ 'of the' (_la couverture des livres_ 'the cover of the books'. At the parsing level, it is better to work only with textforms _des_.

To introduce both textforms and wordforms is a very economic way to encode the two levels of analysis: the text level (needed by parsers) and the linguistic level (useful for linguists who want to study a language).

**Comment by flammie** [comment]
I've used some rule-based pipelines to help our ud-annotation and they usually need to transport this kind of information between modules, for conllu I've so far stuffed this normalised/recased/mangled form in the MISC field, maybe that could be used more widely at least temporarily instead of new field?

**Comment by amir-zeldes** [comment]
I also think  normalization information is quite useful to have. I agree with @flammie that the MISC column is probably the best place to put this right now, since most treebanks will probably not add this information, and since many tools expect 10 columns in conllu, so keeping that stable should be a priority. 

In order to not inflate file sizes unnecessarily, it's possible to only include this annotation in MISC when the normalized value differs from the plain form (which is probably the same in the majority of cases). 

BTW we are also already doing this for Coptic, except in reverse: we always have the normalized form in FORM, and add a MISC annotation Orig=... whenever a text comes from a manuscript with deviating orthography. This makes sense for us because we don't have diplomatic manuscript information for all texts, but we always have normalized data for all words (either because we normalized, or because that's all we had for a text). Here's an example for reference, with Orig indicating the unnormalized form:

https://github.com/UniversalDependencies/UD_Coptic-Scriptorium/blob/dev/cop_scriptorium-ud-dev.conllu#L96

**Comment by foxik** [comment]
What you are talking about is the distinction between **tokens** and **words**, as https://universaldependencies.org/format.html#words-tokens-and-empty-nodes describes them.

While we have a possibility of a token containing multiple words (`Im` -> `in` + `dem`, or `wanna` -> `want` + `to`), for example
```
1-2 wanna
1 want
2 to
```
we do not have a possibility of a **single** token containing a **different** word. The examples where this might be useful:
- original token is not in NFC -> we could have NFC as a word
- punctuation normalization
- spelling errors
- ...

In theory, we could use an obvious extension of multi-word tokens to also single-word tokens as
```
1-1 wich
1 which
```
or even
```
1-1 Hello
1 hello
```
but as far as I remember, people were not very keen to adopt this notion.

Also note that while we are talking about token<->word mapping, we currently have 1->N; single-word tokens are 1->1; and in theory we could also want N->1 (tokens . . . forming a form ...) or even N->M.

**Comment by dan-zeman** [comment]
@foxik it was me who wasn't very keen to adopt "1-1" MWT. I suppose that the word "multi" in multi-word tokens implies > 1 :-) And as I said in #332, if a piece of information pertains to a single surface token, it should appear on the line that is devoted to that token, in the MISC column, unless it is something that according to guidelines belongs to another column.

As for multiple surface tokens that correspond to one syntactic word, UD already uses different mechanisms to capture them:

- if they don't contain a whitespace character, only a hyphen or something, simply adjust tokenization and say it is one token
- if there is a whitespace, connect the two tokens via specialized relations such as `goeswith`, `fixed`, `flat`, `compound`
- in exceptional and well-defined cases, a "word with spaces" may be approved for the given language

**Comment by foxik** [comment]
Yeah, exactly; after our discussion, we plan to represent 1-1 in UDPipe exactly as @amir-zeldes and you suggest, using some key in the MISC field (I am not very keen on just `Orig`, I was thinking more like `OriginalToken`).

Yes, the spaces in tokens allow the M-1 mapping from tokens to words (so tokens "10" and "000" can form word "10 000" :-) I am not suggesting we should provide some explicit M-1 mapping (it would be difficult in CoNLL-U), but we were brainstorming about some text representation lately and ended up with talking about explicit token-word correspondence and what subset does CoNLL-U allow (compared for example to what is allowed in MRP or other semantic parsing formats).
